File: Middle/src/it/dani/selfhome/model/Device.java
package it.dani.selfhome.model;

import org.jetbrains.annotations.NotNull;

public class Device implements Comparable<Device> {
    private final String name;
    private DeviceState state;

    public Device(@NotNull String name)
    {
        this.name = name;
        this.state = new DeviceState((short)0);
    }

    public String getName()
    {
        return this.name;
    }

    public DeviceState getState()
    {
        return this.state;
    }

    public void setState(DeviceState state)
    {
        this.state = (state == null)? new DeviceState((short)0) : state;
    }

    @Override
    public String toString()
    {
        return this.name + " -- " + this.state;
    }

    @Override
    public boolean equals(Object other)
    {
        boolean isEq = this.getClass().equals(other.getClass());
        Device otherD = null;
        if(isEq) otherD = (Device)other;
        return isEq && this.name.equals(otherD.name);
    }

    @Override
    public int compareTo(Device o) {
        return this.name.compareTo(o.name);
    }
}

File: Middle/src/main/kotlin/it/dani/model/Device.kt
package it.dani.model

import it.dani.selfhome.DeviceStateChangeException

data class Device(
    val name: String,
    val initDeviceState: DeviceState
) {
    var onSetState: (DeviceState) -> Any = { }
    val onStateChange: MutableList<(Device) -> Any> = ArrayList()
    var onGetState: (Device) -> DeviceState = { this.initDeviceState }

    var onSetBrightness: (DeviceState) -> Any = { }
    val onBrightnessChange: MutableList<(Device) -> Any> = ArrayList()
    var onGetBrightness: (Device) -> DeviceState = { this.initDeviceState }

    val settings: MutableMap<String, String> = HashMap()

    var state: DeviceState = this.initDeviceState
        set(value) {
            try {
                this.onSetState(value)
                field = value
                this.onStateChange.forEach { it(this) }
            } catch (e: DeviceStateChangeException) {
                throw e
            }
        }
        get() {
            val value = this.onGetState(this)

            if(field != value) {
                field = value
                this.onStateChange.forEach { it(this) }
            }

            return field
        }

    var brightness: DeviceState = this.initDeviceState
        set(value) {
            if(field != value) {
                try {
                    this.onSetBrightness(value)
                    field = value
                    this.onBrightnessChange.forEach { it(this) }
                } catch (e: DeviceStateChangeException) {
                    throw e
                }
            }
        }
        get() {
            val value = this.onGetBrightness(this)

            if(field != value) {
                field = value
                this.onBrightnessChange.forEach { it(this) }
            }

            return field
        }

    override fun toString(): String {
        return "$name [$state] [$brightness] $settings"
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is Device) return false

        if (name != other.name) return false
        if (settings != other.settings) return false

        return true
    }

    override fun hashCode(): Int {
        var result = name.hashCode()
        result = 31 * result + settings.hashCode()
        return result
    }
}